### PR TITLE
add: pallet clockin

### PR DESCRIPTION
**Pull request by ruibinx**
- add: init clockin pallet
- add: clock in meta
- add: clock in

{Summary}

Fixes #{issue}

**Type of change**

* [ ] Bug Fix (non-breaking change which fixes an issue)
* [ ] New Feature (non-breaking change which adds functionality)
* [ ] Breaking Change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update
* [ ] This change requires a storage migration
* [ ] This change requires a benchmark

**Checklist**

* [ ] My code follows the style guidelines of this project
* [ ] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas
* [ ] I have made corresponding changes to the documentation
* [ ] My changes generate no new warnings
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] New and existing unit tests pass locally with my changes
* [ ] New and existing benchmarks pass locally with my changes
* [ ] Any dependent changes have been merged and published in downstream modules
* [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: pallets/clockin/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight(0)]
+        pub fn enable_clock_in(
+            origin: OriginFor<T>,
+            nft_id: NftOf<T>,
+            payout_base: BalanceOf<T>,
+            payout_min: BalanceOf<T>,
+            payout_max: BalanceOf<T>,
+            metadata: Vec<u8>,
```
Comment (coderfengyun) [inline pallets/clockin/src/lib.rs:116]: 这里的metadata代表什么？

File: pallets/clockin/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+            Ok(())
+        }
+
+        #[pallet::weight(0)]
+        pub fn clock_in(origin: OriginFor<T>, nft_id: NftOf<T>) -> DispatchResult {
+            let (did, who) = parami_did::EnsureDid::<T>::ensure_origin(origin)?;
+            let meta = Metadata::<T>::get(nft_id).ok_or(Error::<T>::ClockInNotExists)?;
+
+            let current_height = <frame_system::Pallet<T>>::block_number();
+            let last_clock_in_bucket: HeightOf<T> = LastClockIn::<T>::get(nft_id, did).into();
```
Comment (coderfengyun) [inline pallets/clockin/src/lib.rs]: 如之前讨论这样的设计无法将导致用户级设计无法删除。能否设计为：只是查看用户在当前的bucket是否领过？

File: pallets/clockin/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+        NotNftOwner,
+        NftNotMinted,
+        InsufficientToken,
+        ClockInNotExists,
+        ClockedIn,
+        NumberConversionError,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
```
Comment (coderfengyun) [inline pallets/clockin/src/lib.rs:108]: 逻辑中能否将于bucket相关逻辑提取出来，与纯粹业务分离开？

File: pallets/clockin/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+            ensure!(free_balance >= 0u32.into(), Error::<T>::InsufficientToken);
+
+            let reward = reward.min(free_balance);
+            T::Assets::transfer(meta.asset_id, &meta.pot, &who, reward, false)?;
+
+            let clock_in_bucket = (current_height - meta.start_at) / meta.bucket_size;
+            let clock_in_bucket: u32 = clock_in_bucket
+                .try_into()
+                .map_err(|_| Error::<T>::NumberConversionError)?;
+            let clock_in_bucket = clock_in_bucket + 1;
```
Comment (coderfengyun) [inline pallets/clockin/src/lib.rs]: +1为了解决什么问题？

File: pallets/clockin/src/tests.rs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+        let pot = meta.pot;
+        assert_eq!(Assets::balance(nft_id, pot), 50);
+
+        let mut tag_count = 0;
+        for _ in TagsOf::<Test>::iter_prefix_values(nft_id) {
+            tag_count += 1;
+        }
+        assert_eq!(tag_count, 3);
+        assert_ok!(ClockIn::add_token_reward(Origin::signed(ALICE), nft_id, 49));
```
Comment (coderfengyun) [inline pallets/clockin/src/tests.rs]: 这个没必要了，下面的case测到了

File: pallets/clockin/src/tests.rs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        let mut tag_count = 0;
+        for _ in TagsOf::<Test>::iter_prefix_values(nft_id) {
+            tag_count += 1;
+        }
+        assert_eq!(tag_count, 3);
+    });
+}
+
+#[test]
+fn should_disable_clockin() {
```
Comment (coderfengyun) [inline pallets/clockin/src/tests.rs:135]: disable时需要验证钱还给了nft owner

File: pallets/clockin/src/tests.rs [hunk 7]
```diff
@@ -0,0 +1,170 @@
+use crate::mock::*;
```
Comment (coderfengyun) [inline pallets/clockin/src/tests.rs:1]: 需要增加对于pot是否会冲突的测试

File: pallets/clockin/src/lib.rs (hunk 2 above)
Comment (ruibinx) [reply pallets/clockin/src/lib.rs]: 只是查看用户在当前的bucket是否领过： 这也是一个用户级的存储?而且因为bucket不是连续的，更无法删除

File: pallets/clockin/src/lib.rs (hunk 1 above)
Comment (ruibinx) [reply pallets/clockin/src/lib.rs:116]: 同ad的metadata，是一个外部链接

File: pallets/clockin/src/lib.rs (hunk 4 above)
Comment (ruibinx) [reply pallets/clockin/src/lib.rs]: 计算clock_in_bucket的时候除法是向下取整的，所以要加1。

例如: start_at是0, bucket_size是10，上次打卡的block是11，block至少要到 (11 / 10 + 1) * 10 + 0 = 20, 才可以再次打卡。

File: pallets/clockin/src/lib.rs (hunk 2 above)
Comment (coderfengyun) [reply pallets/clockin/src/lib.rs]: 如果存储的设计是bucket为最前前缀，那么可以直接remove_prefix ?

File: pallets/clockin/src/lib.rs (hunk 4 above)
Comment (coderfengyun) [reply pallets/clockin/src/lib.rs]: 0 - 9，第一次打卡；10 - 19，第二次打卡；
看起来也没有问题？没看明白+1的必要性

File: pallets/clockin/src/tests.rs (hunk 5 above)
Comment (ruibinx) [reply pallets/clockin/src/tests.rs]: done

File: pallets/clockin/src/tests.rs (hunk 6 above)
Comment (ruibinx) [reply pallets/clockin/src/tests.rs:135]: done

File: pallets/clockin/src/tests.rs (hunk 7 above)
Comment (ruibinx) [reply pallets/clockin/src/tests.rs:1]: done

File: pallets/clockin/src/lib.rs (hunk 4 above)
Comment (ruibinx) [reply pallets/clockin/src/lib.rs]: +1 加的是bucket，不是block，不加一的话，在11打卡，算出来的block number 是 10，+1后算出的是20

File: pallets/clockin/src/lib.rs (hunk 3 above)
Comment (ruibinx) [reply pallets/clockin/src/lib.rs:108]: done

File: pallets/clockin/src/lib.rs (hunk 2 above)
Comment (ruibinx) [reply pallets/clockin/src/lib.rs]: 按照nft_id也可以remove_prefix

File: pallets/clockin/src/tests.rs [hunk 8]
```diff
@@ -168,3 +171,124 @@ fn should_disable_clockin() {
...
+            20,
+            b"test".to_vec(),
+            vec![b"tag1".to_vec(), b"tag2".to_vec(), b"tag3".to_vec()],
+            4,
+        ));
+
+        assert_ok!(ClockIn::clock_in(Origin::signed(BOB), nft_id));
+
+        System::set_block_number(20);
+        assert_noop!(
```
Comment (coderfengyun) [inline pallets/clockin/src/tests.rs:289]: 这里不要挂了，直接把剩余的token给了就好

File: pallets/clockin/src/lib.rs (hunk 2 above)
Comment (coderfengyun) [reply pallets/clockin/src/lib.rs]: 也可以

**Review by coderfengyun — COMMENTED** [review]
还差一个查询nft是否有clockin活动的state call接口吧？

File: pallets/clockin/src/tests.rs (hunk 8 above)
Comment (ruibinx) [reply pallets/clockin/src/tests.rs:289]: 只有剩余金额为0的时候会抛异常

**Review by coderfengyun — APPROVED** [review]

**Review by coderfengyun — APPROVED** [review]

**Pull request merged**